File: python/frameworks/langchain/agent.py
"""
LangChain agent with Scalekit-authenticated Gmail tools.
Run: python python/frameworks/langchain/agent.py
"""
import os
import scalekit.client
from dotenv import load_dotenv, find_dotenv
from langchain_core.messages import HumanMessage, ToolMessage
from langchain_openai import ChatOpenAI

load_dotenv(find_dotenv())

scalekit_client = scalekit.client.ScalekitClient(
    client_id=os.getenv("SCALEKIT_CLIENT_ID"),
    client_secret=os.getenv("SCALEKIT_CLIENT_SECRET"),
    env_url=os.getenv("SCALEKIT_ENVIRONMENT_URL"),
)
actions = scalekit_client.actions

response = actions.get_or_create_connected_account(
    connection_name="gmail",
    identifier="user_123",
)
if response.connected_account.status != "ACTIVE":
    link = actions.get_authorization_link(connection_name="gmail", identifier="user_123")
    print("Authorize Gmail:", link.link)
    input("Press Enter after authorizing...")

tools = actions.langchain.get_tools(
    identifier="user_123",
    connection_names=["gmail"],
)
tool_map = {t.name: t for t in tools}

llm = ChatOpenAI(model=os.getenv("LITELLM_MODEL", "claude-sonnet-4-6")).bind_tools(tools)
messages = [HumanMessage("Fetch my last 5 unread emails and summarize them")]

while True:
    response = llm.invoke(messages)
    messages.append(response)
    if not response.tool_calls:
        print(response.content)
        break
    for tool_call in response.tool_calls:
        result = tool_map[tool_call["name"]].invoke(tool_call["args"])
        messages.append(ToolMessage(content=str(result), tool_call_id=tool_call["id"]))

File: javascript/frameworks/vercel-ai/agent.ts
/**
 * Vercel AI SDK agent with Scalekit-authenticated Gmail tools.
 * Run: npx tsx javascript/frameworks/vercel-ai/agent.ts
 */
import { ScalekitClient } from '@scalekit-sdk/node'
import { ConnectorStatus } from '@scalekit-sdk/node/lib/pkg/grpc/scalekit/v1/connected_accounts/connected_accounts_pb.js'
import { createOpenAICompatible } from '@ai-sdk/openai-compatible'
import { generateText, jsonSchema, stepCountIs, tool } from 'ai'
import 'dotenv/config'

const scalekit = new ScalekitClient(
  process.env.SCALEKIT_ENVIRONMENT_URL!,
  process.env.SCALEKIT_CLIENT_ID!,
  process.env.SCALEKIT_CLIENT_SECRET!,
)

const { connectedAccount } = await scalekit.actions.getOrCreateConnectedAccount({
  connectionName: 'gmail',
  identifier: 'user_123',
})
if (connectedAccount?.status !== ConnectorStatus.ACTIVE) {
  const { link } = await scalekit.actions.getAuthorizationLink({
    connectionName: 'gmail',
    identifier: 'user_123',
  })
  console.log('Authorize Gmail:', link)
  process.exit(0)
}

const { tools: scopedTools } = await scalekit.tools.listScopedTools('user_123', {
  filter: { connectionNames: ['gmail'] },
})

const createTool = tool as unknown as (config: Record<string, unknown>) => unknown

const tools = Object.fromEntries(
  scopedTools
    .map((scopedTool) => scopedTool.tool?.definition)
    .filter((definition): definition is NonNullable<typeof definition> => Boolean(definition?.name))
    .map((definition) => [
      String(definition.name),
      createTool({
        description: String(definition.description ?? ''),
        parameters: jsonSchema((definition.input_schema ?? { type: 'object', properties: {} }) as never),
        execute: async (args: Record<string, unknown>) => {
          const result = await scalekit.actions.executeTool({
            toolName: String(definition.name),
            identifier: 'user_123',
            toolInput: args,
          })
          return result.data
        },
      }),
    ]),
)

const provider = createOpenAICompatible({
  name: 'litellm',
  baseURL: process.env.OPENAI_BASE_URL!,
  apiKey: process.env.OPENAI_API_KEY!,
})

const { text } = await generateText({
  model: provider(process.env.OPENAI_MODEL || 'claude-sonnet-4-6'),
  tools: tools as never,
  stopWhen: stepCountIs(5) as never,
  prompt: 'Fetch my last 5 unread emails and summarize them',
})
console.log(text)
